src/backend/distributed/executor/repartition.c

```diff
@@ -0,0 +1,169 @@
+static void TraverseJobTree(Job *curJob, List **jobs);
+static char * GenerateCreateSchemasCommand(List *jobIds);
+static char * GenerateJobCommands(List *jobIds, char *templateCommand);
+static char * GenerateDeleteJobsCommand(List *jobIds);
+
+
+/*
+ * ExecuteDependentTasks executes all tasks except the top level tasks
+ * in order from the task tree. At a time, it can execute different tasks from
+ * different jobs.
+ */
+List *
+ExecuteDependentTasks(List *topLevelTasks, Job *topLevelJob)
+{
+	EnsureNoModificationsHaveBeenDone();
+
+	List *allTasks = TaskAndExecutionList(topLevelTasks);
+
+	List *jobIds = CreateTemporarySchemasForMergeTasks(topLevelJob);
+
+	ExecuteTasksInDependencyOrder(allTasks, topLevelTasks);
+
+	return jobIds;
+}
+
+
+/*
+ * CreateTemporarySchemasForMergeTasks creates the necessary schemas that will be used
+ * later in each worker. Single transaction is used to create the schemas.
+ */
+static List *
+CreateTemporarySchemasForMergeTasks(Job *topLeveLJob)
+{
+	List *jobIds = ExtractJobsInJobTree(topLeveLJob);
+	char *createSchemasCommand = GenerateCreateSchemasCommand(jobIds);
+	SendCommandToAllWorkers(createSchemasCommand);
+	return jobIds;
+}
+
+
+/*
+ * ExtractJobsInJobTree returns all job ids in the job tree
+ * where the given job is root.
+ */
+static List *
+ExtractJobsInJobTree(Job *job)
+{
+	List *jobIds = NIL;
+	TraverseJobTree(job, &jobIds);
+	return jobIds;
+}
+
+
+/*
+ * TraverseJobTree does a dfs in the current job and adds
+ * all of its job ids.
+ */
+static void
+TraverseJobTree(Job *curJob, List **jobIds)
+{
+	ListCell *jobCell = NULL;
+	*jobIds = lappend(*jobIds, (void *) curJob->jobId);
+
+	foreach(jobCell, curJob->dependentJobList)
+	{
+		Job *childJob = (Job *) lfirst(jobCell);
+		TraverseJobTree(childJob, jobIds);
+	}
+}
+
+
+/*
+ * GenerateCreateSchemasCommand returns concatanated create schema commands.
+ */
+static char *
+GenerateCreateSchemasCommand(List *jobIds)
+{
+	return GenerateJobCommands(jobIds, WORKER_CREATE_SCHEMA_QUERY);
+}
+
+
+/*
+ * GenerateJobCommands returns concatenated commands with the given template
+ * command for each job id from the given job ids. The returned command is
+ * exactly list_length(jobIds) subcommands.
+ *  E.g create_schema(jobId1); create_schema(jobId2); ...
+ * This way we can send the command in just one latency to a worker.
+ */
+static char *
+GenerateJobCommands(List *jobIds, char *templateCommand)
+{
+	StringInfo createSchemaCommand = makeStringInfo();
+	ListCell *jobIdCell = NULL;
+
+	foreach(jobIdCell, jobIds)
+	{
+		uint64 jobId = (uint64) lfirst(jobIdCell);
+		appendStringInfo(createSchemaCommand, templateCommand, jobId);
+	}
+	return createSchemaCommand->data;
+}
+
+
+/*
+ * DoRepartitionCleanup removes the temporary job directories and schemas that are
+ * used for repartition queries for the given job ids.
+ */
+void
+DoRepartitionCleanup(List *jobIds)
+{
+	SendCommandListToAllWorkers(list_make1(GenerateDeleteJobsCommand(jobIds)));
```

Comment:
We need to make sure not to generate any errors here, since it's called from a PG_CATCH. `SendCommandListToAllWorkers` may throw errors iirc.

Reply:
you mean we should have a pg_catch inside SendCommandListToAllWorkers  as well?

Reply:
we should have an "optional" variant, e.g. like SendOptionalCommandListToWorkerInTransaction

Reply:
is this possibly what you meant https://github.com/citusdata/citus/pull/3169/commits/3dfc24f7a80bbcc8c7a7672f603016fe18c44b7b